runtime/sak.c

```diff
@@ -0,0 +1,103 @@
+/*                                                                        */
+/*                                 OCaml                                  */
+/*                                                                        */
+/*                 David Allsopp, OCaml Labs, Cambridge.                  */
+/*                                                                        */
+/*   Copyright 2021 David Allsopp Ltd.                                    */
+/*                                                                        */
+/*   All rights reserved.  This file is distributed under the terms of    */
+/*   the GNU Lesser General Public License version 2.1, with the          */
+/*   special exception on linking described in the file LICENSE.          */
+/*                                                                        */
+/**************************************************************************/
+
+/* Runtime Builder's Swiss Army Knife. This utility performs functions
+   previously delegated to classic Unix utilities but which ultimately seem to
+   cause more hassle for maintenance than the initial simplicity suggests.
+
+   This tool is a memorial to the many hours and PRs spent chasing down strange
+   locale issues, stray CR characters and fighting yet another incompatible
+   implementation of sed or awk. */
+
+/* Borrow the Unicode *_os definitions and T() macro from misc.h */
+#define CAML_INTERNALS
+#include "caml/misc.h"
+
+#include <stdio.h>
+#include <string.h>
+
+/* Operations
+   - translate. Used for the OCAML_STDLIB_DIR macro in runtime/build_config.h
+     to ensure the LIBDIR make variable is correctly represented as a C string
+     literal.
+
+     On Unix, `sak translate /usr/local/lib` returns `"/usr/local/lib"`
+
+     On Windows, `sak translate "C:\OCaml🐫\lib"` returns
+     `L"C:\\OCaml\xd83d\xdc2b\\lib"`
+ */
+
+void usage(void)
+{
+  printf(
+    "OCaml Build System Swiss Army Knife\n"
+    "Usage: sak command\n"
+    "Commands:\n"
+    " * translate path - converts path to a C string constant\n"
+  );
+}
+
+/* Converts the supplied path (UTF-8 on Unix and UCS-2ish on Windows) to a valid
+   C string literal. On Windows, this is always a wchar_t* (L"..."). */
+void emit_c_string(char_os *path)
+{
+  char_os c;
+
+#ifdef _WIN32
+  putchar('L');
+#endif
+  putchar('"');
+
+  while ((c = *path++) != 0) {
+    /* Escape \, " and \n */
+    if (c == '\\') {
+      printf("\\\\");
+    } else if (c == '"') {
+      printf("\\\"");
+    } else if (c == '\n') {
+      printf("\\n");
+#ifndef _WIN32
+    /* On Unix, nothing else needs escaping */
+    } else {
+      putchar(c);
+#else
+    /* On Windows, allow 7-bit printable characters to be displayed literally
+       and escape everything else (using the older \x notation for increased
+       compatibility, rather than the newer \U. */
+    } else if (c < 0x80 && iswprint(c)) {
```

Comment:
Is the condtion `iswprint` necessary? The documentation said that this function is locale-sensitive, could that be an issue?

Reply:
The locale part is "safe", and is the reason for rolling our own C program vs shell utilities. C _requires_ that programs start in the C locale, so in `runtime/sak` we don't have to worry about the environment settings (`sed`, `sort`, etc. call `setlocale(LC_ALL, "")` to inherit the environment's setting).

This doesn't have to be done using `iswprint` - it's equivalent to:

```suggestion
    } else if (c > 0x1F && c < 0x7F) {
```
if you'd prefer?

Reply:
The function `iswprint` itself is fine. I was just wondering why escaping the control character is necessary on Windows.

Reply:
Paranoia more than anything else - the lower control characters (0x00-0x1F) should never be seen because they're not valid in filenames. I just don't trust that _all_ `cl`s which are out there would correctly read 0x1F (which is valid in a filename, if tricky to type in the Windows shell) in a source string!